### PR TITLE
Problem: memiavl is not integrated with state machine

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #938 

Solution:
- implement custom rootmulti store and kvstore wrappers

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: memiavl/import.go [hunk 1]
```diff
@@ -1,16 +1,129 @@
...
+		} else if err != nil {
+			return snapshottypes.SnapshotItem{}, errors.Wrap(err, "invalid protobuf message")
+		}
+
+		switch item := snapshotItem.Item.(type) {
+		case *snapshottypes.SnapshotItem_Store:
+			if importer != nil {
+				importer.Close()
+			}
+			importer := NewTreeImporter(filepath.Join(snapshotDir, item.Store.Name), int64(height))
```
Comment (mmsqe) [inline memiavl/import.go:48]:
```suggestion
			importer = NewTreeImporter(filepath.Join(snapshotDir, item.Store.Name), int64(height))
```

File: memiavl/db.go [hunk 2]
```diff
@@ -335,5 +343,9 @@ func initEmptyDB(dir string, initialVersion uint32) error {
 	if err := tmp.WriteSnapshot(snapshotDir); err != nil {
 		return err
 	}
-	return os.Symlink(snapshotDir, currentPath(dir))
+	tmpPath := currentTmpPath(dir)
+	if err := os.Symlink(filepath.Base(snapshotDir), tmpPath); err != nil {
+		return err
+	}
+	return os.Rename(tmpPath, currentPath(dir))
```
[the remark this answers is not in the document]
Comment (yihuang) [reply memiavl/db.go:350]: great, done.

File: memiavl/db.go [hunk 3]
```diff
@@ -256,7 +254,11 @@
 	}
 
 	db.MultiTree = *mtree
-	db.pendingUpgrades = nil
+
+	if len(db.pendingUpgrades) > 0 {
+		db.MultiTree.ApplyUpgrades(db.pendingUpgrades)
```
[the remark this answers is not in the document]
Comment (mmsqe) [reply memiavl/db.go:257]: might need handle this error

File: memiavl/export.go [hunk 4]
```diff
@@ -1,9 +1,61 @@
...
 	"github.com/cosmos/iavl"
+	protoio "github.com/gogo/protobuf/io"
 )
 
+func (db *DB) Snapshot(height uint64, protoWriter protoio.Writer) error {
+	if height > math.MaxUint32 {
+		return fmt.Errorf("height overflows uint32: %d", height)
+	}
+
+	mtree, err := LoadMultiTree(snapshotPath(db.dir, uint32(height)))
```
[the remark this answers is not in the document]
Comment (yihuang) [reply memiavl/export.go:18]: added error wrapping for this.

**Review by mmsqe — APPROVED** [review]

**Pull request merged**